### DISCUSSION TITLE
Are we correctly using the icon for the password visibility action?

### DISCUSSION THREAD

**Discussion by dgdavid** [agama-project/agama#766] [General]
Before going into details, please make the effort to look at the picture below as if you were in front of this kind of form for the first time ever. 

![Screenshot 2023-09-24 at 23-18-31 Agama](https://github.com/openSUSE/agama/assets/1691872/2f47a625-18e7-481d-8d1a-1e04b7e9469e)

Look at the open, not crossed-out eye icon and answer sincerely: what is it telling you? 

* a) the password input is in the "show" state, displaying introduced characters
* b) the password input is in the "hide" state, obscuring introduced characters

While [I requested](https://github.com/openSUSE/agama/pull/750#pullrequestreview-1627588514) the use of this eye pattern in the code review, I found myself giving the first answer when doing manual testing. It was like, _"Wait, is the dialog opening with the password inputs in the show state?"_.

Instead of opening the discussion in the PR itself and putting it at risk of getting blocked for a while, I decided to take notes and open a discussion here because there are a lot of web, desktop, and even mobile applications out there using the same "eye pattern". Without going any further, the Cockpit's login page uses it as we do. And we often don't question things that are widely used.

However, I think worth it to rethink the whole thing a bit. In the worst case (to me) we could end up keeping it as it is.

The first idea that came to my mind was proposing to use the icon the other way around: the crossed out variant when the input is obscuring the password and the other when not. But you could argue that it might be a bad idea to go against the grain. And you would be right: We shouldn't use a wide use "icon action" with a different meaning than the rest of the world... even though they are all wrong :stuck_out_tongue_winking_eye: 

So, as an alternative and [similar to the first approach](https://github-production-user-asset-6210df.s3.amazonaws.com/112288843/267802797-a19f5efd-5d77-455a-b3da-7ddd0334c208.png) sent by @balsa-asanovic when developing the feature, we could use a plain text action. I.e., to use the "Show" and "Hide" texts instead of icons. But we will be in the same dilemma: is the text representing the current input state, or is the action to be performed when clicking on it?

But, if the dilemma is about displaying the input status or the next action, could we solve it by displaying a visual hint in sync with the current status... :thinking: :thinking:  An old plain [HTML checkbox](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox) with its label? A [PatternFly Switch](https://www.patternfly.org/components/switch)?

|Inline, long label version | Inline, short label version | In the next _line_ |
|-|-|-|
|![Screenshot 2023-09-23 at 19-48-48 Agama](https://github.com/openSUSE/agama/assets/1691872/19e87090-0f7a-42bc-8a3d-03bd5db4b668) | ![Screenshot 2023-09-24 at 19-37-05 Agama](https://github.com/openSUSE/agama/assets/1691872/d61ffa6b-de76-426b-8436-a4c207bac4e2) | ![Screenshot 2023-09-24 at 19-40-38 Agama](https://github.com/openSUSE/agama/assets/1691872/6eb4d829-6b71-4495-a46f-bd391c095baa) |

<details>
<summary>Click to show/hide a rough diff</summary>

---

```diff
diff --git a/web/src/assets/styles/patternfly-overrides.scss b/web/src/assets/styles/patternfly-overrides.scss
index 6b43b8ac..2e4ad2a4 100644
--- a/web/src/assets/styles/patternfly-overrides.scss
+++ b/web/src/assets/styles/patternfly-overrides.scss
@@ -186,7 +186,8 @@ table td > .pf-v5-c-empty-state {
   }
 }

-// Center icon in the visibility button of password input form fields
-.password-toggler span.pf-c-button__icon {
-  display: flex;
+// Let's InputGroupText be as wide as its content (motivated by Show Password
+// switch)
+.pf-v5-c-input-group__item.pf-m-box {
+  min-inline-size: fit-content;
 }
diff --git a/web/src/components/core/PasswordInput.jsx b/web/src/components/core/PasswordInput.jsx
index 14a55ab5..b7452ea2 100644
--- a/web/src/components/core/PasswordInput.jsx
+++ b/web/src/components/core/PasswordInput.jsx
@@ -30,16 +30,15 @@
  */
 import React, { useState } from "react";
 import {
-  Button,
   InputGroup,
-  TextInput
+  InputGroupText,
+  TextInput,
+  Switch
 } from "@patternfly/react-core";
-import { Icon } from "~/components/layout";
 import { _ } from "~/i18n";

 export default function PasswordInput({ id, ...props }) {
   const [showPassword, setShowPassword] = useState(false);
-  const visibilityIconName = showPassword ? "visibility_off" : "visibility";

   if (!id) {
     const field = props.label || props["aria-label"] || props.name;
@@ -53,15 +52,15 @@ export default function PasswordInput({ id, ...props }) {
         id={id}
         type={showPassword ? 'text' : 'password'}
       />
-      <Button
-        id={`toggle-${id}-visibility`}
-        className="password-toggler"
-        aria-label={_("Password visibility button")}
-        variant="control"
-        onClick={() => setShowPassword((prev) => !prev)}
-        icon={<Icon name={visibilityIconName} size="15" />}
-        isDisabled={props.isDisabled}
-      />
+      <InputGroupText>
+        <Switch
+          id={`toggle-${id}-visibility`}
+          label={_("Show password")}
+          isChecked={showPassword}
+          onChange={() => setShowPassword((prev) => !prev)}
+          isDisabled={props.isDisabled}
+        />
+      </InputGroupText>
     </InputGroup>
   );
 }
```
</details>

<em>Examples above inspired by https://joshwayne.com/posts/show-password/ article and a Friday afternoon chat with @joseivanlopez</em>

---

Fun fact,

Do you know what? While doing a quick research (actually when writing this question) I found something interesting: a very big company [was using the icon in the way I was about to suggest above](https://ux.stackexchange.com/questions/123812/is-google-using-the-password-show-hide-toggle-icon-on-their-login-form-correctly), but if you go to its [login form](https://github.com/openSUSE/agama/assets/1691872/5bc6dd55-791d-4f94-98f5-83873d61e1e1) right now you will see a plain HTML "Show password" checkbox there  (aka "Mostrar contraseña") . Which might mean something.

**Comment by joseivanlopez** [comment]
You see, Google already realized everyone was using it wrongly. And not to mention the password confirmation, but that is for another discussion :).

**Comment by dgdavid** [comment]
> but that is for another discussion :)

Already started :roll_eyes:  https://github.com/openSUSE/agama/discussions/767 

I fear dropping it is not so straightforward for our use case.

**Comment by balsa-asanovic** [comment]
> Look at the open, not crossed-out eye icon and answer sincerely: what is it telling you?
>
>    a) the password input is in the "show" state, displaying introduced characters
>    b) the password input is in the "hide" state, obscuring introduced characters

Indeed, when looking at it with blank slate, the eye would intuitively indicate that you can see the password, while the crossed eye would be telling me that I can't.

> And you would be right: We shouldn't use a wide use "icon action" with a different meaning than the rest of the world... even though they are all wrong 😜

Haha, agreed.

I personally like the approach you suggested, using switch/checkbox, since this makes it quite clear that if the switch is ON, the password should be visible, and the other way around.

I prefer the "Inline, short label version" and after it the "In the next line", while I like the least "Inline, long label version", though this is purely based on some subjective sense of taste, so I could be very wrong 😄 .

If it is decided to go on with this change, I'd of course be willing to work on it. However, you did already provide most of the code. 😅

**Reply by dgdavid** [reply]
> I prefer the "Inline, short label version" and after it the "In the next line"

I had exactly the same preference as you, but the _weird_ thing with the `PF/InputGroup#aria-describedby` attribute and the thing with the tab order (see https://github.com/openSUSE/agama/discussions/765) made me change my mind. Now I'd go with the "[ ] Show password" node in the _next line_ whenever we manage to make it the last thing of the form.

But this is not about my preference :wink: That's why I opened the discussion.

Thanks a lot for your insights, much appreciated. Really.

**Reply by dgdavid** [reply]
> If it is decided to go on with this change, I'd of course be willing to work on it. 

:rocket: :tada: 

> However, you did already provide most of the code. 😅

Nah, just an example. There is work to do, like adjusting it properly and writing/doing some tests. And if you are willing to do so... :wink:

But first, let's see what's the conclusion of this discussion. We have to wait a bit for gathering more feedback / ideas.

**Reply by balsa-asanovic** [reply]
Of course, let's see where the discussion takes us.

**Comment by dgdavid** [comment]
Almost _by accident_, today I've landed at https://github.com/alphagov/govuk-design-system-backlog/issues/240 and read their [interesting rationale](https://github.com/alphagov/govuk-design-system-backlog/issues/240#issuecomment-2020125340) about stop using the confirmation input which [we decided to keep](https://github.com/agama-project/agama/discussions/767).

Additionally, I have read https://github.com/alphagov/govuk-design-system/issues/2795 which could help us either, to re-evaluate our decision to keep the confirmation input :innocent: or to finally improve the "Show" password actions. Lately I was thinking about combine them in just one and move it after the confirmation for improving the keyboard interaction. It's quite annoying to me to get trapped in the "Show" action when I wanted to go for the confirmation. But, as you can see in the previous link, they recommend still using two actions but with different labels to reduce the chances of confusing the user. Still, I wonder if would be enough one action properly labeled. E.g., "Show entered passwords", "Show password and password confirmation", or something similar.

Last but not least, following other links found in the mentioned issue, I've realize about the security risk about these "Show password" actions (https://www.bleepingcomputer.com/news/security/google-microsoft-can-get-your-passwords-via-web-browsers-spellcheck/) and the possible and simple solution (https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/spellcheck#security_and_privacy_concerns)

So, a lot of things for feeding a PBI :wink:

**Reply by grisu48** [reply]
Additional input from a tester.

I think the current state of allowing users to input their root password without any confirmation doesn't give enough weight to the importance of the root password. It is surely annoying to type in a strong root password twice, but it's much more annoying to end up being locked out of your system afterwards.

As someone who was (and still to some extend is) system administrator, I do operate very much with muscle memory when it comes to the root password. Having me show the password is useless because the password lives in my muscles not in my eyesight.

Also I see that the audience for this kind of user input is completely different than the target audience for the articles above. Most of them address password fields in the context of web browsers, where password managers have become an established standard. In this context it makes really no sense to ask for confirmation. Passwords in this context are also not considered safe without additional MFA, a situation that is different for the root password. Recovery options are also readily available.

***

I think for a system installer that targets IT professional looking at the needs of web technology for the general public gives us misleading ideas about the usability. Something as important as the root password certainly requires confirmation in a way that suits the user. Not having confirmation here I would argue is unacceptable and having confirmation via visible feedback does not suit the user basis that operates out of muscle memory like I do.

**Reply by grisu48** [reply]
TLDR: I would argue that the password confirmation text input field is the better choice for the target audience.

**Reply by grisu48** [reply]
Another aspect is accessibility. For blind/impaired people it might be easier to write the password twice than to look at it again. This might be even impossible for some people.

**Reply by dgdavid** [reply]
Interesting POV, thanks a lot for the input.

**Reply by grisu48** [reply]
Suggestion: Ask for password confirmation by default but allow the user to skip the confirmation text field by displaying the password.

Something like the screenshot shown in https://github.com/agama-project/agama/discussions/766#discussion-5667252, but when the user displays the password in clear text, they are allowed to proceed without the need to input the password.

This would be an attempt to bring both approaches together.

**Reply by dgdavid** [reply]
Thanks for the suggestion!

Actually, the password confirmation is still there (you can find it under Users > Root Authentication > Password > Change). We've only removed it in the mandatory step for now to give it a try and gather feedback before deciding which one to choose. That mandatory step is already intrusive enough without adding more fields. In fact, you can check out how it was _revamped_ at https://github.com/agama-project/agama/pull/1821

As for your idea of skipping the confirmation when displaying the password, I think it might make the user interaction too complex or unexpected. We’d also need to explain that behavior, which would slightly add more cognitive load.

I really appreciate ideas (especially the crazy ones! :wink:) because they help us learn and improve, but the goal is to keep the interface simple, intuitive, and yet powerful.

That said, we’ll revisit this whole topic in the mid-term, according to priorities and collected feedback (like yours).

Thanks again. Keep going!

**Reply by dgdavid** [reply]
> That mandatory step is already intrusive enough

Which, by the way, is intended to be improved as well. But this is a bit off topic for this thread and we've no estimation about how and when yet, since we're now focused in other stuff.

**Reply by grisu48** [reply]
Thank you, I'm looking forward to see those ideas evolving :+1:

**Comment by dgdavid** [comment]
At some point in time, I believe we are going to simply drop the icon and go for a more explicit `[ ] Show password` option which should be clear for everyone.
